xla/service/gpu/cudnn_norm_rewriter.cc

```diff
@@ -930,23 +959,32 @@ class CudnnNormRewriterVisitor : public DfsHloRewriteVisitor {
       }
 
       // Verify the broadcasts of scale and bias.
-      if (!ShapeUtil::EqualIgnoringElementType(reduce->operand(0)->shape(),
-                                               broadcast_scale->shape()) ||
-          !ShapeUtil::EqualIgnoringElementType(reduce->operand(0)->shape(),
-                                               broadcast_bias->shape()) ||
-          reduce->dimensions() != broadcast_scale->dimensions() ||
-          reduce->dimensions() != broadcast_bias->dimensions()) {
+      if (!ShapeUtil::EqualIgnoringElementType(
+              ShapeUtil::DropDegenerateDimensions(reduce->operand(0)->shape()),
+              ShapeUtil::DropDegenerateDimensions(broadcast_scale->shape())) ||
+          !ShapeUtil::EqualIgnoringElementType(
+              ShapeUtil::DropDegenerateDimensions(reduce->operand(0)->shape()),
+              ShapeUtil::DropDegenerateDimensions(broadcast_bias->shape())) ||
+          norm_dims_adjusted != AdjustedDimensions(broadcast_scale) ||
+          norm_dims_adjusted != AdjustedDimensions(broadcast_bias)) {
         VLOG(1) << "Layer norm operand broadcast not supported.";
         return absl::OkStatus();
       }
 
       // If necessary, transpose the input so that the dimensions not being
       // normalized are the leading dimensions.
-      std::vector<int64_t> non_norm_dims;
-      for (int64_t x_dim = 0; x_dim < x.Instr()->shape().rank(); ++x_dim) {
+      std::vector<int64_t> non_norm_dims, non_norm_dims_adjusted;
+      for (int64_t x_dim = 0, degen_dims = 0; x_dim < x.Instr()->shape().rank();
+           ++x_dim) {
         if (std::find(norm_dims.begin(), norm_dims.end(), x_dim) ==
             norm_dims.end()) {
           non_norm_dims.emplace_back(x_dim);
+          if (x.Instr()->shape().dimensions(x_dim) != 1) {
+            non_norm_dims_adjusted.emplace_back(x_dim - degen_dims);
```

Comment:
The logic for non_norm_dims_adjusted is a bit hard to follow. Can you create an overload of `AdjustedDimensions` that takes a `Shape` and `Span<int>` of dimensions (with a clear comment describing what it does), have the version taking an instruction call this alternative version, and also call the alternative version here on `non_norm_dims` to get `non_norm_dims_adjusted`?